### PR TITLE
[Fixes #17014] Add an auth-enabled template for Blazor WASM projects

### LINKED ISSUE #17014: Have an auth-enabled Blazor WASM project template

### PR DESCRIPTION
**Pull request by javiercn**
* The template enables auth using cookies.
* Includes a UserController to return the user claims/data to the user
* Includes an IdentityAuthenticationStateProvider that fetches the
  claims data from the application.
* It requires a hosted project.

### CODE REVIEW

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs]: Since the client is agnostic to what technology is being used on the server (i.e., it doesn't care whether it's Identity or something else), could we name this to `ServerAuthenticationStateProvider` or similar?

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs]: Does this need to be public? Could it be a private field?

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs]:
```suggestion
            var user = await Client.GetJsonAsync<UserInfo>("User");
```

... because we don't know whether it's mounted at the root of the domain or not, and the `<base href>` should control this.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs]: Not sure if there's a reason for the way this is structured. Is the following simplification acceptable?

```cs
        public override async Task<AuthenticationState> GetAuthenticationStateAsync()
        {
            if (_currentUser == null)
            {
                await UpdateUser();
            }

            return new AuthenticationState(_currentUser);
}
```

Also I'd have a minor preference for replacing `UpdateUser` with `FetchUser` and having it return a `ClaimsPrincipal`, so then we have:

```cs
_currentUser = await FetchUser();
```

... so that all the concerns about `_currentUser` are scoped to a single method. Not a super big deal though!

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Server/Data/WeatherForecastService.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Server/Data/WeatherForecastService.cs]: Is this used anywhere? I can't see references to it. It seems unnecessary since `WeatherController` takes care of this an is simpler than having a separate "service" concept.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/UserInfo.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/UserInfo.cs]: I'd be interested in restructuring this class a bit to make it clearer that we *don't* expect the server to disclose everything it knows about the `ClaimsPrincipal`. That is,

 * By default we should only expose the "name" claim
 * By default we do not expose `AuthenticationType`, since the client has no use for this info by default. Developers should manually add some code (e.g., the whole `AuthenticationType` property) if they want to communicate this to the client
 * By default we shouldn't expose any other claims, as they might contain sensitive info. We could have a comment telling developers that if they want to, they can write code to populate more of the `Claims` collection using data from `claimsPrincipal`, but it's up to them to decide carefully which claim data is safe to expose to end-users. For example, don't expose a claim if it contains a token that grants access to some other service which is only supposed to be called from the server.

So by default, `UserInfo` might be reducible just to having the `Claims` property, in which by default we only place the "name" claim.

### About roles

I'm slightly less sure about this, but I suspect that similarly we shouldn't assume they want to expose the full list to the client, or even that client-side roles have the same names as server-side ones. We might have a commented-out block of code giving an example of how you can specify a role if you want, e.g.,

```cs
// To pass information about roles to the client-side application, add boolean properties for each client-side
// role and use logic similar to the following to populate them. Note that typically you do not need to expose
// the full list of roles to the client.
// IsSuperAdmin = claimsPrincipal.IsInRole("IsSuperAdmin");
```

**Review by SteveSandersonMS — COMMENTED** [review]
This looks like a great start! I haven't dug through every line of all the complicated templating changes, but it looks totally plausible at first glance. Thanks very much for figuring out all these details. I know it's quite delicate working all this stuff out.

I'll look at that part in more detail later, but for now have posted some initial thoughts on the server/client communication parts.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/UserInfo.cs
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/UserInfo.cs]: > * By default we shouldn't expose any other claims, as they might contain sensitive info.
The default set of claims doesn't include anything sensitive.

> We could have a comment telling developers that if they want to, they can write code to populate more of the `Claims` collection using data from `claimsPrincipal`, but it's up to them to decide carefully which claim data is safe to expose to end-users. 

I think we could have a comment in the other direction saying that if you want to, you can customize the list of claims to share with the user here. Even if you add properties to the IdentityUser they don't get exposed as claims, so I think it is fine.

> For example, don't expose a claim if it contains a token that grants access to some other service which is only supposed to be called from the server.

I don't think this is a valid a valid example. We don't store access tokens inside user claims and our guidance would be not to store tokens in a claim inside the identity. Basically, treat secrets as secrets and don't mix them with any other information.

Also, in this scenario you control the app E2E, so you are only sharing this information with your app.

> ### About roles
> I'm slightly less sure about this, but I suspect that similarly we shouldn't assume they want to expose the full list to the client, or even that client-side roles have the same names as server-side ones. We might have a commented-out block of code giving an example of how you can specify a role if you want, e.g.,

Roles are not secrets.

I think exposing roles enables a better experience by default, as you are able to use them in your pages to limit access directly through the UI to some areas of your app with the appropriate Authorize attribute.

My reasoning behind all these decisions are:
 * You control your app, end to end
 * The authentication implementation is meant to be used with Identity.
 * There is already a place to opt-in to adding claims which defined by Identity (IUserClaimsPrincipalFactory)
    * If you add a claim to your app, normally you want to allow your app to consume it.

**Review by ryanbrandenburg — APPROVED** [review]
Just suggestions, nothing blocking.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Startup.cs
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Startup.cs]: We are not planning to. I’ve kept the current symbols for consistency

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/App.razor
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/App.razor]: This is the same code as in the blazor server-side template.

Keeping it the same for consistency.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/UserInfo.cs
Comment (SteveSandersonMS) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/UserInfo.cs]: This might all be fine, but would it be OK if we raise this in particular when we do security review? My concern is that we're creating a new form of hazard, where in the past the contents of a ClaimsIdentity would not be exposed to the user in plaintext (right?), but now it is. Even if we don't expect someone to put in things that should be secret from the user themselves, them might do so and not realise what's happening here. Examples would be:

 * A role the user should not be allowed to discover they are in (e.g., `HellbannedUsers`)
 * A claim that contains tokens that are only intended for use by the server, not client, even if we don't recommend this

Perhaps these are not valid scenarios but is it OK if we lay this out directly in security review for feedback there? I'm not saying this PR should be blocked on that, since it's only a very small part of this PR and we can tweak it later if needed.

> The authentication implementation is meant to be used with Identity.

I thought our goal was to make it agnostic to the server-side authentication mechanism. There's no reason Blazor WebAssembly should prefer people to use Identity in particular, or any other authentication mechanism. Anything should work as long as it tracks state in cookies.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/.template.config/template.json
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/.template.config/template.json]: We were hoping to expose all the same auth types as are available on the server (None, Individual, IndividualB2C, SingleOrg, MultiOrg, Windows). Having done it for the server template, I know it's super tedious putting in the config for all of these.

Do you think that's viable, or do you anticipate any problems with these?

Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/.template.config/template.json]: Let's take this discussion offline.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/UserInfo.cs
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/UserInfo.cs]: > * A role the user should not be allowed to discover they are in (e.g., `HellbannedUsers`)

I don't think roles are meant to be used that way. Normally they include you in a group, not exclude you from. As in, you have Admin role, you can access this area of the application and so on. 

What is described in this sample doesn't work with the way authorization works in ASP.NET Core either, as the Authorize attribute allows you to require a role, but not to avoid having it.

> I thought our goal was to make it agnostic to the server-side authentication mechanism. There's no reason Blazor WebAssembly should prefer people to use Identity in particular, or any other authentication mechanism. Anything should work as long as it tracks state in cookies.

I think we should have a very clear scope here. The template uses identity and we should not try to build anything more generic than that, based on the assumption that the templates are a starting point and that if you want to replace Identity with your own solution then you have all the power to modify the code in whatever way you see fit.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs]: Do you have any other suggestion? I see this as an extension of Identity on the app if you ask me. I think `ServerAuthenticationStateProvider` is a bad choice given the fact that there is already a `ServerAuthenticationStateProvider` class in ASP.NET Core, but I'm open to suggestions.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs]: Updated.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs]: This was updated, not sure why the diff doesn't show it.

Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs]: Updated, not sure why this doesn't reflect it.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Server/Data/WeatherForecastService.cs
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Server/Data/WeatherForecastService.cs]: Removed

**Comment by javiercn** [comment]
@SteveSandersonMS Other than whether we should enable something other than Individual auth or the naming of the auth provider I believe all feedback is addressed.

Can you take a look?

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/UserInfo.cs
Comment (SteveSandersonMS) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/UserInfo.cs]: This is a bit tangential, but hellbanning doesn't mean exclusion from anything. It's a flag associated with a user meaning that their actions will be treated as silent from the point of view of other users. For example, if you were hellbanned on GitHub, then you could still do all the normal stuff, but when you post comments or PRs, the system would see you were in the "hellbanned" group and flag your post/PR so it's invisible to other users. There are other ways to model this state, but someone *might* reasonably use a role for it without understanding how we're now disclosing that info.

Anyway let's delay the discussion about whether to list all roles until we do security review. We don't need to block this PR on it.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs
Comment (SteveSandersonMS) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs]: Hmm, maybe `HostAuthenticationStateProvider`, since it gets state from the "host" app in the hosted template scenario?

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/.template.config/template.json
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/.template.config/template.json]: This is weird. At first I thought it was a typo, since SQLite is called SQLite (one "L", not two), but then I see we spell it `SqlLite` in all the other templates too.

So I guess I agree we should have the same typo that exists in all the other templates 🤷‍♂

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/ClaimValue.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/ClaimValue.cs]: How would you feel about renaming this namespace (and the containing directory) to `Authorization`?

It would be great to avoid a generic term like `Models` if possible as it doesn't really mean anything here. Pretty much everything in the "shared" project will be some sort of "model class".

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/.template.config/vs-2017.3.host.json
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/.template.config/vs-2017.3.host.json]: Could you clarify how the tooling knows that `Individual` auth is only allowed if the "hosted" parameter is also true?

We discussed this before and I think you mentioned you expected that VS supports imposing this restriction, but I don't spot where it's defined in the config here. If we need to ask the project creation dialog people to add some feature for this, let's start that process ASAP.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/.template.config/template.json
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/.template.config/template.json]: No, it's SQLite, and yes, it also annoys me quite a bit :(.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/ClaimValue.cs
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Models/ClaimValue.cs]: That's fine with me.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs]: The URL specified here needs to be `User`, not `/User`

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/.template.config/vs-2017.3.host.json
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/.template.config/vs-2017.3.host.json]: We might need to, I need to check with Phil

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs]: Yes, I forgot

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs]: That's a pretty strange string. It's just an arbitrary string, right? Could we replace it with something like `nameof(IdentityAuthenticationStateProvider)` or whatever we end up calling the class?

Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Client/Services/IdentityAuthenticationStateProvider.cs]: It will replace it with the project name

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Server/Data/WeatherForecast.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Server/Data/WeatherForecast.cs]: This class isn't used either, right? We have one in the 'shared' project.

Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Server/Data/WeatherForecast.cs]: It is on the controller, unless I'm confused

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Authorization/UserInfo.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Authorization/UserInfo.cs]: What do you think about maybe moving this "populate from claimsPrincipal" logic into the server project (in `UserController`'s `GetCurrentUser`)? This would just be to clarify that the server is the only thing that controls what data gets copied in here - the client doesn't control it, it just receives some serialized data.

As a more minor thing, this logic just isn't needed on the client and won't be stripped out by the linker since the linker doesn't run on user assemblies.

File: src/ProjectTemplates/test/BlazorWasmTemplateTest.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/test/BlazorWasmTemplateTest.cs]: Excellent!

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Server/Data/WeatherForecast.cs
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Server/Data/WeatherForecast.cs]: You're right. I've removed it.

File: src/ProjectTemplates/test/BlazorWasmTemplateTest.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/test/BlazorWasmTemplateTest.cs]: Awesome

**Review by SteveSandersonMS — APPROVED** [review]
Looking great!

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Authorization/UserInfo.cs
Comment (SteveSandersonMS) [inline src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Authorization/UserInfo.cs]: Could the server and client ever have different opinions about `NameClaimType` and `RoleClaimType`?

What I really mean is, if we didn't pass this info, and the client just populated its claimsprincipal with all the claims it's recieved, wouldn't `User.Identity.Name` and the "roles" APIs still work OK? They definitely did work OK like that in my earlier prototypes, but I don't know if there are some extra edge cases you're accounting for here.

Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Authorization/UserInfo.cs]: No, they depend on the NameClaimType and RoleClaimType afaik.

Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Authorization/UserInfo.cs]: It might have worked because you were using default values for both.

File: src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Authorization/UserInfo.cs
Comment (javiercn) [reply src/ProjectTemplates/BlazorWasm.ProjectTemplates/content/BlazorWasm-CSharp/Shared/Authorization/UserInfo.cs]: I've moved this. Not sure why the diff is wrong.

**Pull request merged**
